content/blog/2019-09-25-faster-interpreter.md

```diff
@@ -0,0 +1,154 @@
+
+# Interpreting Bril
+
+[Bril](https://github.com/sampsyo/bril) is a pleasantly simple intermediate representation for
+teaching compilers and compiler optimizations. Part of Bril being useful for this purpose is the
+existence of a solid, simple reference interpreter for the language: `brili`. Written in
+[TypeScript](https://www.typescriptlang.org/), `brili` is straightforward to extend and allows easy
+experimentation with new language features and optimizations. The flipside of `brili`'s simplicity
+and the use of TypeScript for its implementation is that is isn't very fast. Thus, while `brili` is
+suitable for working with small, simple Bril programs, it may make experiments using more complex,
+long-running programs unnecessarily onerous.
+
+# Gotta go fast
+
+Given this, our goal was to make a faster interpreter for Bril. The simplest way to do this is to
+pick a faster language than TypeScript for the interpreter's implementation. As such, we chose to
+[Rewrite It In Rust](https://github.com/ansuz/RIIR). [Rust](https://www.rust-lang.org/) is a
+modern systems programming language designed to make it easy to "build reliable and efficient
+software"; it typically performs comparably to (or sometimes faster than) the same software written
+in a more traditional systems language such as C or C++.
+
+Rust is appealing for this project for a few other reasons. Chief among these is its rich library
+ecosystem, which includes the excellent [Serde](https://serde.rs/). Serde is a library (technically,
+a family of libraries) for **ser**ializing and **de**serializing data between a range of formats,
+including JSON. It provides utilities for automatically [deriving](https://serde.rs/derive.html)
+functions to parse JSON (and other formats) directly into native Rust structs, which is convenient
+for an interpreter. With Serde, we can define the data structures for Bril's various operations and
+overall program structure, then get parsing code from Bril's JSON representation for free! Even
+better, Serde's automatically generated code is typically very fast.
+
+# Interpreter structure
+
+Our first "draft" of an interpreter implemented a simple structure:
+1. Using Serde, derive JSON deserialization code for types representing the Bril operations
+and other syntax (including labels, functions, etc.)
+
+2. Parse Bril JSON into the aforementioned structures.
+
+3. Form basic blocks and a CFG from the program instructions. This process constructs an mapping
+from label names to basic block indices and uses this map to "link" together connected basic blocks.
+Linking in this way (rather than just constructing the name/index mapping and using that at runtime)
+lets us have slightly faster execution - it's faster to load an array index than it is to e.g.
+load a heap-allocated object (as in a pointer-linked graph) or compute a string hash every time we
+need to find a label location. It's worth noting that while we could've gotten away with not forming
```

Comment:
It should be slightly faster in common cases due to cache locality, I think? Many Bril programs have fairly short basic blocks; the basic block representation is also pretty small (a vector of instructions and a vector of exit indices; instructions are at most a vector of `usize`s (8 bytes each) + a few more bytes of overhead). Given that, it's reasonable that many Bril programs could fit in cache (at least L2 or L3; maybe L1 for some very small programs). So, keeping blocks in a vector and using indices lets us take advantage of contiguous allocation - and therefore cache locality - more so than we could with pointers.

If Rust made it easy to store pointers to vector elements, we could do that instead and get the same benefit a bit more ergonomically. However, as far as I know, this is a pain because of the borrow checker.

Does that seem reasonable/right?

Reply:
Sure; yes, now I see! You’re storing the blocks continuously in memory, which gets you spatial locality. Thanks for clarifying!

Reply:
I'll add a polished version of that explanation to the blog post; if the speed gain was unclear to you, it'll probably be unclear to a broader audience too.

Reply:
Added in b43fba69

Reply:
Excellent!